File: 3396-minimum-number-of-operations-to-make-elements-in-array-distinct/3396-minimum-number-of-operations-to-make-elements-in-array-distinct.py
class Solution:
    def minimumOperations(self, nums: List[int]) -> int:
        ans, counter, i, n = 0, Counter(nums), 0, len(nums)

        while True:
            unique_count = 0
            for key, val in counter.items():
                if val == 1:
                    unique_count += 1
            
            if unique_count == len(counter):
                break

            for _ in range(min(3, n - i)):
                counter[nums[i]] -= 1
                if counter[nums[i]] == 0:
                    del counter[nums[i]]
                i += 1

            ans += 1

        return ans

File: 3396-minimum-number-of-operations-to-make-elements-in-array-distinct/3396-minimum-number-of-operations-to-make-elements-in-array-distinct.cpp
class Solution {
public:
    int minimumOperations(vector<int>& nums) {
        int ans = 0, i = 0, n = nums.size();
        unordered_map<int, int> freq;
        for (int num : nums) ++freq[num];

        while (true){
            int uniqueCount = 0;

            for (auto& [key, val]: freq){
                if (val == 1) ++uniqueCount;
            }

            if (uniqueCount == freq.size()) break;
            
            int left = n - i;
            for (int j = 0; j < min(3, left); ++j){
                --freq[nums[i]];
                
                if (freq[nums[i]] == 0) freq.erase(nums[i]);
                ++i;
            }

            ++ans;
        }

        return ans;
    }
};